### PR TITLE
[Safetensors] Add explicit  flag to from pretrained

### PR DESCRIPTION
**Pull request by patrickvonplaten**
## Give user more control when loading safetensors

It would be very helpful if we could give the user more control over loading safetensors checkpoints by adding a `use_safetensors` flag to `from_pretrained`. 

At the moment, `safetensors` weights are **always** loaded when `safetensors` is installed and are **silently** not loaded if `safetensors` is installed, **but** the model has no `safetensors` weights. 

By giving the user the option to do `use_safetensors=True/False` we enable two new use cases:

1.) If a user only wants to load models from safetensors checkpoints, they can now pass `use_safetensors=True` which will lead to an error if no safetensors checkpoints are available => this can then e.g. give the user the guarantee that no pickle is used when loading checkpoints

2.) If a user doesn't want to load `safetensors` checkpoints, but not uninstall the library, they can now pass `use_safetensors=False` which will never load safetensors checkpoitns. This is super helpful for testing as well.

Also this feature would unblock this PR: in `diffusers`. https://github.com/huggingface/diffusers/pull/2123

### CODE REVIEW

**Review by amyeroberts — APPROVED** [review]
LGTM! 

I'm not super familiar with the weights loading code so would be good to have the opinions of @sgugger and @Narsil

**Review by sgugger — APPROVED** [review]
Thanks for adding this!

**Comment by patrickvonplaten** [comment]
Should I add a test before merging or just merge?

File: @ [hunk 1]
```diff
@@ -0,0 +1,3538 @@
+# coding=utf-8
```
Comment (sgugger) [inline @]: I don't think you meant to add this file.

File: tests/test_modeling_common.py [hunk 2]
```diff
@@ -3327,6 +3329,49 @@ def test_legacy_load_from_url(self):
...
+            in str(env_error.exception)
+        )
+
+        # test that error if only safetensors is available
+        with self.assertRaises(OSError) as env_error:
+            BertModel.from_pretrained("hf-internal-testing/tiny-random-bert-safetensors", use_safetensors=False)
+
+        self.assertTrue("does not appear to have a file named pytorch_model.bin" in str(env_error.exception))
+
+        # test that only safetensors if both available if set to True
```
Comment (sgugger) [inline tests/test_modeling_common.py:3349]:
```suggestion
        # test that only safetensors if both available and use_safetensors=False
```
it's the opposite ;-)

File: tests/test_modeling_common.py [hunk 3]
```diff
@@ -3327,6 +3329,49 @@ def test_legacy_load_from_url(self):
...
+                subfolder="text_encoder",
+                use_safetensors=False,
+                cache_dir=tmp_dir,
+            )
+
+            all_downloaded_files = glob.glob(os.path.join(tmp_dir, "*", "snapshots", "*", "*", "*"))
+            self.assertTrue(any(f.endswith("bin") for f in all_downloaded_files))
+            self.assertFalse(any(f.endswith("safetensors") for f in all_downloaded_files))
+
+        # test that no safetensors if both available if set to False
```
Comment (sgugger) [inline tests/test_modeling_common.py:3362]:
```suggestion
        # test that no safetensors if both available and use_safetensors=True
```

**Review by sgugger — COMMENTED** [review]
Nice new test, careful not to merge with the added file!

File: @ (hunk 1 above)
Comment (patrickvonplaten) [reply @]: yeah already removed :-)

**Pull request merged**